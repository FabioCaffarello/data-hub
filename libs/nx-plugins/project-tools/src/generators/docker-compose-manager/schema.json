{
  "$schema": "http://json-schema.org/schema",
  "cli": "nx",
  "$id": "DockerComposeManager",
  "title": "",
  "type": "object",
  "properties": {
    "groupServices": {
      "type": "array",
      "items": {
        "type": "string"
      },
      "description": "Select the group services to include in your compose",
      "x-prompt": {
        "type": "select",
        "message": "Select the group services to include in your compose",
        "items": [
          { "value": "bot", "label": "Bot" },
          { "value": "back-end", "label": "Back-end" },
          { "value": "front-end", "label": "front-end" }
        ],
        "multiple": true
      }
    },
    "coreBusiness": {
      "type": "string",
      "description": "coreBusiness"
    },
    "subjectBusiness": {
      "type": "string",
      "description": "subjectBusiness"
    },
    "ingestors": {
      "type": "string",
      "description": "list the bots name"
    }
  },
  "required": ["groupServices"]
}
