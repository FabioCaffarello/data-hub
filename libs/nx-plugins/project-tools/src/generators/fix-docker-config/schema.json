{
  "$schema": "http://json-schema.org/schema",
  "cli": "nx",
  "$id": "FixDockerConfig",
  "title": "",
  "type": "object",
  "properties": {
  },
  "required": []
}
