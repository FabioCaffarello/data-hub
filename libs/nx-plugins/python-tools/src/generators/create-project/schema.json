{
  "$schema": "http://json-schema.org/schema",
  "cli": "nx",
  "$id": "NxPluginsCreateProject",
  "title": "",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use?"
    },
    "type": {
      "type": "string",
      "enum": ["application", "library"],
      "description": "Project type application or library",
      "default": "application",
      "x-prompt": {
        "message": "What project type would you like to use?",
        "type": "list",
        "items": [
          {
            "value": "application",
            "label": "Application"
          },
          {
            "value": "library",
            "label": "Library"
          }
        ]
      }
    },
    "packageName": {
      "type": "string",
      "description": "Python package name",
      "x-prompt": "What package name would you like to use?"
    },
    "moduleName": {
      "type": "string",
      "description": "Python module name"
    },
    "description": {
      "type": "string",
      "description": "Project short description",
      "x-prompt": "What description would you like to use?"
    },
    "tags": {
      "type": "string",
      "description": "Add tags to the project (used for linting)",
      "alias": "t"
    },
    "directory": {
      "type": "string",
      "description": "A directory where the project is placed"
    }
  },
  "required": ["name", "type"]
}
