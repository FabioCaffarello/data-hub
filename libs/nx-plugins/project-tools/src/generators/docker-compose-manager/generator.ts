import {
  formatFiles,
  generateFiles,
  ProjectConfiguration,
  Tree
} from '@nrwl/devkit';
import { readWorkspaceConfig } from '@nrwl/workspace';
import * as fs from 'fs';
import * as path from 'path';
import * as yaml from 'yaml';
import { DockerComposeManagerGeneratorSchema } from './schema';
import { propsMappings, serviceProps, stackMappings } from './stack.services';
import { stackTemplates, networkName } from './template.services';
import { getOwnKeys } from '../../../../utils/json/commands.json'


interface NormalizedSchema extends DockerComposeManagerGeneratorSchema {
  botService: boolean;
  backendService: boolean;
  frontendService: boolean;
  ingestorsObject: {botName: string, botPath: string}[];
  workspacePath: string;
  dockerComposePath: string;
}


const devContainerJsonPath = path.join('.devcontainer', 'devcontainer.json');


function normalizeOptions(tree: Tree, options: DockerComposeManagerGeneratorSchema): NormalizedSchema {
  const { groupServices } = options;
  const workspacePath = getWorkspacePath(tree)
  const dockerComposePath = path.join(workspacePath, 'docker-compose.yml');
  let ingestorsObject: {botName: string, botPath: string}[]

  for (const group of groupServices) {
    const prop = propsMappings[group];
    if (prop) {
      serviceProps[prop] = true;
    } else {
      console.warn(`Unknown group '${group}'`);
    }
  }

  if (serviceProps.botService) {
    ingestorsObject = (options.ingestors ?? '')
    .split(',')
    .map((name) => ({
      botName: `ingestors-${options.coreBusiness}-${options.subjectBusiness}-${name}`,
      botPath: getProjectPath(tree, `ingestors-${options.coreBusiness}-${options.subjectBusiness}-${name}`),
    }));
  }

  return {
    ...options,
    ...serviceProps,
    ingestorsObject,
    workspacePath,
    dockerComposePath
  };
}

function getProjectPath(tree: Tree, projectName: string): string {
  const workspaceConfig = readWorkspaceConfig({
    format: 'nx',
    path: path.join(tree.root, 'workspace.json'),
  });
  const projectConfig: ProjectConfiguration | undefined =
    workspaceConfig.projects[projectName];

  return projectConfig?.root ?? '';
}

function addFiles(tree: Tree, options: NormalizedSchema) {
  const templateOptions = {
    ...options,
    offsetFromRoot: '',
    template: ''
  };
  generateFiles(
    tree,
    path.join(__dirname, 'files'),
    options.workspacePath,
    templateOptions
  );
}

function getWorkspacePath(tree: Tree): string {
  const devContainerJson = JSON.parse(fs.readFileSync(devContainerJsonPath, 'utf-8'));
  const workspaceFolder = devContainerJson['workspaceFolder'];
  if (fs.existsSync(devContainerJsonPath)) {
    if (workspaceFolder.includes(tree.root)) {
      return path.join(tree.root, '..', '..')
    } else {
      return tree.root
    }
  } else {
    return ''
  }
}

function addServices(tree: Tree, options: NormalizedSchema) {
  const dockerComposeContent = tree.read(options.dockerComposePath, 'utf-8');
  const dockerComposeObject = dockerComposeContent
    ? yaml.parse(dockerComposeContent)
    : {};
  const dockerComposeServices = dockerComposeObject.services || {};

  if (options.botService || options.backendService || options.frontendService) {
    // All stacks
    for (const stack of getOwnKeys(stackMappings)) {
      stackMappings[stack].forEach((element) => {
        const stackTemplatesObject = { ...stackTemplates[element] };
        dockerComposeServices[element] = stackTemplatesObject;
      });
    }
    if (options.botService) {
      // Bot Stack
      options.ingestorsObject.forEach((element) => {
        const botServiceObject = {
          build: {
            context: element.botPath,
            dockerfile: 'Dockerfile' // path.join(element.botPath, 'Dockerfile')
          },
          networks: [networkName]
        };
        dockerComposeServices[element.botName] = botServiceObject;
      })
    }
  }
  dockerComposeObject.services = dockerComposeServices;
  const updatedDockerComposeContent = yaml.stringify(dockerComposeObject);
  tree.write(options.dockerComposePath, updatedDockerComposeContent);
}

export default async function (tree: Tree, options: DockerComposeManagerGeneratorSchema) {
  const normalizedOptions = normalizeOptions(tree, options);
  addFiles(tree, normalizedOptions);
  addServices(tree, normalizedOptions);
  await formatFiles(tree);
}
