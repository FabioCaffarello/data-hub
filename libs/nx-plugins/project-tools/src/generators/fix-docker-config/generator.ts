import {
  Tree,
} from '@nrwl/devkit';
import { readJson, updateJson } from '../../../../utils/json/commands.json';
import { FixDockerConfigGeneratorSchema } from './schema';
import { expect } from '@jest/globals';


const filePath = '../.docker/config.json'

export default async function (tree: Tree, options: FixDockerConfigGeneratorSchema) {

  const configData = readJson(filePath)
  delete configData.credsStore;
  updateJson(filePath, configData);
}

export function assertGenerateFiles(tree: Tree) {
  const config = readJson(filePath);
  expect(config.credsStore).toBeUndefined();
}
