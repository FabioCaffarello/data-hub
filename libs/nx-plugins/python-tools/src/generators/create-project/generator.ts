import {
  addProjectConfiguration,
  formatFiles,
  generateFiles,
  getWorkspaceLayout,
  names,
  offsetFromRoot,
  Tree,
} from '@nrwl/devkit';
import * as path from 'path';
import { PyprojectToml } from '@nxlv/python'
import { parse, stringify } from '@iarna/toml';
import { CreateProjectGeneratorSchema } from './schema';
import { getCurrentGitUserName, getCurrentGitUserEmail } from '../../../../utils/git/commands.git'
// import spawn from 'cross-spawn';
import { spawnSync } from 'child_process';


interface NormalizedSchema extends CreateProjectGeneratorSchema {
  projectName: string;
  projectRoot: string;
  projectDirectory: string;
  parsedTags: string[];
  moduleName: string;
  authorName: string;
  authorEmail: string;
}

function normalizeOptions(tree: Tree, options: CreateProjectGeneratorSchema): NormalizedSchema {
  const name = names(options.name).fileName;
  const projectDirectory = options.directory
    ? `${names(options.directory).fileName}/${name}`
    : name;
  const projectName = projectDirectory.replace(new RegExp('/', 'g'), '-');
  const moduleName = options.moduleName ? options.moduleName : projectName.replace(new RegExp('-', 'g'), '_');

  let projectRoot = '';
  if (options.type === 'application') {
    projectRoot = `${getWorkspaceLayout(tree).appsDir}/${projectDirectory}`;
  } else {
    projectRoot = `${getWorkspaceLayout(tree).libsDir}/${projectDirectory}`;
  }
  const parsedTags = options.tags
    ? options.tags.split(',').map((s) => s.trim())
    : [];

  const authorName = getCurrentGitUserName()
  const authorEmail = getCurrentGitUserEmail()

  return {
    ...options,
    description: options.description ?? '',
    projectName,
    projectRoot,
    projectDirectory,
    parsedTags,
    moduleName,
    authorName,
    authorEmail,
  };
}

function addFiles(tree: Tree, options: NormalizedSchema) {
    const templateOptions = {
      ...options,
      ...names(options.name),
      offsetFromRoot: offsetFromRoot(options.projectRoot),
      template: ''
    };
    generateFiles(tree, path.join(__dirname, 'files'), options.projectRoot, templateOptions);
}

function updateRootPyprojectToml(
  tree: Tree,
  normalizedOptions: NormalizedSchema
) {
  if (tree.exists('./pyproject.toml')) {
    const rootPyprojectToml = parse(
      tree.read('pyproject.toml', 'utf-8')
    ) as PyprojectToml;
    rootPyprojectToml.tool.poetry.dependencies[normalizedOptions.packageName] =
      {
        path: normalizedOptions.projectRoot,
        develop: true,
      };
      tree.write('pyproject.toml', stringify(rootPyprojectToml));
  }
}

function updateRootPoetryLock(tree: Tree, normalizedOptions: NormalizedSchema) {
  if (tree.exists('./pyproject.toml')) {
    console.log(`Updating root poetry.lock...`);
    const executable = 'poetry';
    const updateArgs = ['update', normalizedOptions.packageName];
    spawnSync(executable, updateArgs, {
      shell: false,
      stdio: 'inherit',
    });
    return
  }
}

function installSharedPythonCore(normalizedOptions: NormalizedSchema) {
  console.log(
    `Add shared python core to the project ${normalizedOptions.projectName}`
  );
  const executable = 'npx';
  const installArgs = [
    'nx',
    'run',
    `${normalizedOptions.projectName}:add`,
    '--name',
    'shared-python-tools-core',
    '--local',
  ];
  spawnSync(executable, installArgs, {
    shell: false,
    stdio: 'inherit',
  });
  return
}

function installSharedPythonDevelopment(normalizedOptions: NormalizedSchema) {
  console.log(
    `Add shared python development to the project ${normalizedOptions.projectName}`
  );
  const executable = 'npx';
  const installArgs = [
    'nx',
    'run',
    `${normalizedOptions.projectName}:add`,
    '--name',
    'shared-python-tools-development',
    '--local',
    '--group',
    'dev',
  ];
  spawnSync(executable, installArgs, {
    shell: false,
    stdio: 'inherit',
  });
  return
}

async function generator(tree: Tree, options: CreateProjectGeneratorSchema) {
  const normalizedOptions = normalizeOptions(tree, options);
  addProjectConfiguration(
    tree,
    normalizedOptions.projectName,
    {
      root: normalizedOptions.projectRoot,
      projectType: normalizedOptions.type,
      sourceRoot: `${normalizedOptions.projectRoot}/${normalizedOptions.moduleName}`,
      targets: {
        lock: {
          executor: '@nrwl/workspace:run-commands',
          options: {
            command: 'poetry lock --no-update',
            cwd: normalizedOptions.projectRoot,
          },
        },
        add: {
          executor: '@nxlv/python:add',
          options: {},
        },
        update: {
          executor: '@nxlv/python:update',
          options: {},
        },
        remove: {
          executor: '@nxlv/python:remove',
          options: {},
        },
        install: {
          executor: '@nxlv/python:install',
          options: {
            silent: false,
            args: '',
            cacheDir: `.cache/pypoetry`,
            verbose: false,
            debug: false,
          },
        }
      },
      tags: normalizedOptions.parsedTags,
    }
  );
  addFiles(tree, normalizedOptions);
  updateRootPyprojectToml(tree, normalizedOptions);
  await formatFiles(tree);

  return () => {
    installSharedPythonCore(normalizedOptions);
    installSharedPythonDevelopment(normalizedOptions);
    updateRootPoetryLock(tree, normalizedOptions);
  };
}

export default generator;
