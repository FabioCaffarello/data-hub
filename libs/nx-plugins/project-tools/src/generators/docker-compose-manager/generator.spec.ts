import {
  Tree,
} from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import generator from './generator';
import { DockerComposeManagerGeneratorSchema } from './schema';

describe('docker-compose-manager generator', () => {
  let appTree: Tree;
  const options: DockerComposeManagerGeneratorSchema = {
    groupServices: ['bot'],
    coreBusiness: 'financial',
    subjectBusiness: 'mercadobitcoin',
    ingestors: 'tickers',
  };

  beforeEach(() => {
    appTree = createTreeWithEmptyWorkspace({layout: 'apps-libs'});
  });

  it('should generate a docker-compose file', async () => {
    await generator(appTree, options);
    assertGenerateFiles(appTree);

  });
});

function assertGenerateFiles(
  appTree: Tree,
) {
  expect(appTree.exists(`${appTree.root}/docker-compose.yml`)).toBeTruthy();
  expect(
    appTree.read(`${appTree.root}/docker-compose.yml`).toString()
  ).toMatchSnapshot();
}
