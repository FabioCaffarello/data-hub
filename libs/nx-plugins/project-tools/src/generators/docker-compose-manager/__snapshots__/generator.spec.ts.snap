// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`docker-compose-manager generator should generate a docker-compose file 1`] = `
"version: \\"3\\"
services:
  mongodb:
    image: mongo
    ports:
      - 27017:27017
    volumes:
      - ./data:/data/db
    networks:
      - data-hub-network
  rabbitmq:
    image: rabbitmq:3-management
    ports:
      - 5672:5672
      - 15672:15672
    environment:
      RABBITMQ_DEFAULT_USER: guest
      RABBITMQ_DEFAULT_PASS: guest
      RABBITMQ_DEFAULT_VHOST: /
    networks:
      - data-hub-network
  ingestors-financial-mercadobitcoin-tickers:
    build:
      context: apps/ingestors/financial/mercadobitcoin/tickers
      dockerfile: apps/ingestors/financial/mercadobitcoin/tickers/Dockerfile
    networks:
      - data-hub-network
networks:
  data-hub-network:
    external: true
    driver: bridge
"
`;
